### PR TITLE
ServerConfig accepts an sdf::Root DOM object

### PR DESCRIPTION
**Pull request by nkoenig**
Signed-off-by: Nate Koenig <nate@openrobotics.org>

# 🎉 New feature

## Summary

Requires: https://github.com/ignitionrobotics/sdformat/pull/841

I have a downstream program that is using Gazebo as a library. This program runs a world with a number of models. Each model can have different starting conditions, which is mostly just pose information for now but can be expanded to joint positions and forces.

A key aspect is consistent and reproducible behavior. To meet this goal, I would like to start Gazebo with all models at their specified locations, and not rely on asynchronous spawning. This goal can be accomplished by sending the `Server` a complete SDF DOM.

I have modified the `ServerConfig` to accept an `sdf::Root` object as an alternative to the file and string options.

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [x] Updated documentation (as needed)
- [x] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [x] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

**Comment by nkoenig** [comment]
The `findFuelResourceSdf` function in Util.cc is a direct copy from Server.cc, and the new `resolveSdfWorldFile` function is a copy of the code removed from `Server.cc`'s constructor.

**Comment by mjcarroll** [comment]
Also, not for gazebo6, since it would screw up API, but for main I think we could remove the "SdfString" and "SdfFile" accessors/mutators and make ServerConfig users always parse the Sdf into a root before construction the server config.

**Review by mjcarroll — COMMENTED** [review]
Two nits and a question

File: include/ignition/gazebo/ServerConfig.hh [hunk 1]
```diff
@@ -175,6 +176,17 @@ namespace ignition
...
+      /// \brief Set the SDF Root DOM object. The sdf::Root object will take
+      /// precendence over ServerConfig::SdfString() and
+      /// ServerConfig::SdfFile().
+      /// \param[in] _root SDF Root object to use.
+      public: void SetSdfRoot(const sdf::Root &_root) const;
+
+      /// \brief Get the SDF Root DOM object.
+      /// \return SDF Root object to use, or std::nullopt if the sdf::Root
+      /// has not been set via ServerConfig::SetSdfRoot().
+      public: const std::optional<sdf::Root> &SdfRoot() const;
```
[the remark this answers is not in the document]
Comment (nkoenig) [reply include/ignition/gazebo/ServerConfig.hh]: Changed to non-const in [f32555b](https://github.com/ignitionrobotics/ign-gazebo/pull/1333/commits/f32555b2884167d7da28613229f447dbfaace483)

File: include/ignition/gazebo/Util.hh [hunk 2]
```diff
@@ -210,6 +210,24 @@ namespace ignition
...
+    /// \brief Convert an SDF world filename string, such as "shapes.sdf", to
+    /// full system file path.
+    /// The provided SDF filename may be a Fuel URI, relative path, name
+    /// of an installed Gazebo world filename, or an absolute path.
+    /// \param[in] _sdfFile An SDF world filename such as:
+    ///    1. "shapes.sdf" - This is referencing an installed world file.
+    ///    2. "../shapes.sdf" - This is referencing a relative world file.
+    ///    3. "/home/user/shapes.sdf" - This is reference an absolute world
+    ///       file.
+    ///    4. "https://fuel.ignitionrobotics.org/1.0/openrobotics/worlds/shapes.sdf" - This is referencing a Fuel URI. This will download the world file.
```
[the remark this answers is not in the document]
Comment (nkoenig) [reply include/ignition/gazebo/Util.hh]: Wrapped in [f32555b](https://github.com/ignitionrobotics/ign-gazebo/pull/1333/commits/f32555b2884167d7da28613229f447dbfaace483)

File: src/Util_TEST.cc [hunk 3]
```diff
@@ -37,6 +39,7 @@
 #include "ignition/gazebo/Util.hh"
 
 #include "helpers/EnvTestFixture.hh"
+#include "helpers/UniqueTestDirectoryEnv.hh"
```
[the remark this answers is not in the document]
Comment (nkoenig) [reply src/Util_TEST.cc]: Removed in [f32555b](https://github.com/ignitionrobotics/ign-gazebo/pull/1333/commits/f32555b2884167d7da28613229f447dbfaace483)

**Comment by nkoenig** [comment]
> Since we are expanding the number of ways that sdf can be injected into the server config, do you think it would make sense to introduce an enum like:
> 
> ```
> enum class SdfSource {
>   kSdfString,
>   kSdfFile,
>   kSdfRoot
> }
> ```
> 
> To make it easier to check and make switch statements around the branching logic?

Added the enum in [f32555b](https://github.com/ignitionrobotics/ign-gazebo/pull/1333/commits/f32555b2884167d7da28613229f447dbfaace483)

File: include/ignition/gazebo/ServerConfig.hh [hunk 4]
```diff
@@ -36,6 +37,22 @@ namespace ignition
...
     class ServerConfigPrivate;
 
+    /// \brief Type of SDF source.
+    enum class SourceType
+    {
+      // The source is an SDF file.
+      kSdfFile,
+
+      // No source specified.
+      kNone,
```
Comment (mjcarroll) [inline include/ignition/gazebo/ServerConfig.hh:47]: Is there a specific reason that kNone isn't first?

File: include/ignition/gazebo/ServerConfig.hh [hunk 5]
```diff
@@ -36,6 +37,22 @@ namespace ignition
     // Forward declarations.
     class ServerConfigPrivate;
 
+    /// \brief Type of SDF source.
+    enum class SourceType
```
Comment (mjcarroll) [inline include/ignition/gazebo/ServerConfig.hh:41]: I think this should either be inside `ServerConfig` or have a more descriptive name so we don't accidentally clash inside the `ignition::gazebo` namespace in the future.

File: src/ServerConfig_TEST.cc [hunk 6]
```diff
@@ -228,3 +228,27 @@ TEST(ServerConfig, GenerateRecordPlugin)
...
+  config.SetSdfFile("file");
+  EXPECT_FALSE(config.SdfRoot());
+  EXPECT_FALSE(config.SdfFile().empty());
+  EXPECT_TRUE(config.SdfString().empty());
+
+  sdf::Root root;
+  config.SetSdfRoot(root);
+  EXPECT_TRUE(config.SdfRoot());
+  EXPECT_TRUE(config.SdfFile().empty());
+  EXPECT_TRUE(config.SdfString().empty());
```
Comment (mjcarroll) [inline src/ServerConfig_TEST.cc:253]: Add assertions for SourceType

File: src/Util.cc [hunk 7]
```diff
@@ -579,6 +582,80 @@ std::optional<math::Vector3d> sphericalCoordinates(Entity _entity,
...
+    {
+      return current;
+    }
+  }
+  return "";
+}
+
+//////////////////////////////////////////////////
+std::string IGNITION_GAZEBO_VISIBLE resolveSdfWorldFile(
+    const std::string &_sdfFile, const std::string &_fuelResourceCache)
```
Comment (mjcarroll) [inline src/Util.cc:613]: I think this is tripping up the linter

```suggestion
std::string resolveSdfWorldFile(
    const std::string &_sdfFile, const std::string &_fuelResourceCache)
```

**Review by mjcarroll — CHANGES_REQUESTED** [review]

File: include/ignition/gazebo/ServerConfig.hh (hunk 4 above)
Comment (nkoenig) [reply include/ignition/gazebo/ServerConfig.hh:47]: I was going alphabetic order. I'll move it to first.

File: src/Util.cc (hunk 7 above)
Comment (nkoenig) [reply src/Util.cc]: Fixed in [7d33cf0](https://github.com/ignitionrobotics/ign-gazebo/pull/1333/commits/7d33cf0aafacefa3a4bdaa1522ce02b7d7b5263b)

File: src/ServerConfig_TEST.cc (hunk 6 above)
Comment (nkoenig) [reply src/ServerConfig_TEST.cc:256]: Added in [7d33cf0](https://github.com/ignitionrobotics/ign-gazebo/pull/1333/commits/7d33cf0aafacefa3a4bdaa1522ce02b7d7b5263b)

File: include/ignition/gazebo/ServerConfig.hh (hunk 5 above)
Comment (nkoenig) [reply include/ignition/gazebo/ServerConfig.hh]: Moved to be inside `ServerConfig` in [7d33cf0](https://github.com/ignitionrobotics/ign-gazebo/pull/1333/commits/7d33cf0aafacefa3a4bdaa1522ce02b7d7b5263b)

File: include/ignition/gazebo/ServerConfig.hh (hunk 4 above)
Comment (nkoenig) [reply include/ignition/gazebo/ServerConfig.hh]: Changed in [7d33cf0](https://github.com/ignitionrobotics/ign-gazebo/pull/1333/commits/7d33cf0aafacefa3a4bdaa1522ce02b7d7b5263b)

**Review by mjcarroll — APPROVED** [review]

**Comment by nkoenig** [comment]
Needs a release of sdformat12.

**Comment by ahcorde** [comment]
@osrf-jenkins retest this please

**Comment by nkoenig** [comment]
@osrf-jenkins rerun tests please

**Comment by nkoenig** [comment]
> 
I love magic

**Comment by nkoenig** [comment]
That also explains why github actions didn't start when I triggered a rerun.

**Comment by chapulina** [comment]
I manually clicked retry on the Jenkins builds. GitHub Actions seems to be down too.

Here's where the incantation is documented, for future reference:

https://ignitionrobotics.org/docs/all/ci#triggering-ci

**Comment by nkoenig** [comment]
@osrf-jenkins run tests

**Review by chapulina — APPROVED** [review]
LGTM, just some small nits about documentation

**Comment by chapulina** [comment]
Merging it with my documentation comments not addressed, they're minor.

**Comment by osrf-triage** [comment]
This pull request has been mentioned on **Gazebo Community**. There might be relevant details there:

https://community.gazebosim.org/t/new-ignition-releases-2022-04-13-fortress-edifice/1367/1

**Pull request merged**